x/accounts/defaults/multisig/account.go

```diff
@@ -0,0 +1,341 @@
+	_, err = a.Votes.Get(ctx, collections.Join(msg.ProposalId, sender))
+	if err == nil && !cfg.Revote {
+		return nil, errors.New("voter has already voted, can't change its vote per config")
+	}
+	if err != nil && !errors.Is(err, collections.ErrNotFound) {
+		return nil, err
+	}
+
+	addr, err := a.addrCodec.BytesToString(sender)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = a.eventService.EventManager(ctx).EmitKV("vote",
+		event.NewAttribute("proposal_id", fmt.Sprint(msg.ProposalId)),
+		event.NewAttribute("voter", addr),
+		event.NewAttribute("vote", msg.Vote.String()),
+	); err != nil {
+		return nil, err
+	}
+
+	return &v1.MsgVoteResponse{}, a.Votes.Set(ctx, collections.Join(msg.ProposalId, sender), int32(msg.Vote))
+}
+
+func (a Account) CreateProposal(ctx context.Context, msg *v1.MsgCreateProposal) (*v1.MsgCreateProposalResponse, error) {
+	// check if the sender is a member
+	_, err := a.Members.Get(ctx, accountstd.Sender(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	seq, err := a.Sequence.Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// check if the proposal already exists
+	_, err = a.Proposals.Get(ctx, seq)
+	if err == nil {
+		return nil, errors.New("proposal already exists")
+	}
+
+	config, err := a.Config.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// create the proposal
+	proposal := v1.Proposal{
+		Title:           msg.Proposal.Title,
+		Summary:         msg.Proposal.Summary,
+		Messages:        msg.Proposal.Messages,
+		Execute:         msg.Proposal.Execute,
+		VotingPeriodEnd: a.hs.HeaderInfo(ctx).Time.Add(time.Second * time.Duration(config.VotingPeriod)).Unix(),
+		Status:          v1.ProposalStatus_PROPOSAL_STATUS_VOTING_PERIOD,
+	}
+
+	if err = a.Proposals.Set(ctx, seq, proposal); err != nil {
+		return nil, err
+	}
+
+	addr, err := a.addrCodec.BytesToString(accountstd.Sender(ctx))
+	if err != nil {
+		return nil, err
+	}
+
+	if err = a.eventService.EventManager(ctx).EmitKV("proposal_created",
+		event.NewAttribute("proposal_id", fmt.Sprint(seq)),
+		event.NewAttribute("proposer", addr),
+	); err != nil {
+		return nil, err
+	}
+
+	return &v1.MsgCreateProposalResponse{ProposalId: seq}, nil
+}
+
+func (a Account) ExecuteProposal(ctx context.Context, msg *v1.MsgExecuteProposal) (*v1.MsgExecuteProposalResponse, error) {
+	prop, err := a.Proposals.Get(ctx, msg.ProposalId)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := a.Config.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// check if voting period is still active and early execution is disabled
+	if a.hs.HeaderInfo(ctx).Time.Unix() < prop.VotingPeriodEnd && !config.EarlyExecution {
+		return nil, errors.New("voting period has not ended yet")
+	}
+
+	// perform tally
+	rng := collections.NewPrefixedPairRange[uint64, []byte](msg.ProposalId)
+	yesVotes := uint64(0)
+	noVotes := uint64(0)
+	abstainVotes := uint64(0)
+	err = a.Votes.Walk(ctx, rng, func(key collections.Pair[uint64, []byte], vote int32) (stop bool, err error) {
```

Comment:
how come we need to iterate votes here? we should do lazy tallying instead of counting it all at the end imo.

Reply:
What happens if voting weights change in the middle of a vote? So there are proposals A and B:

- A changes the voting weight of Facu from 30 to 0, will be executed in 1h
- B is a random proposal.

Facu votes on B before A is executed, so then when proposal B passes it would have accounted for a member that doesn't exist anymore.

A solution would be that when a power changes, we check if the changed member voted in any active proposals and change the vote accordingly.

```
Proposal 1
yes: 100
no: 50

Facu's weight changed from 60 to 40, and he voted yes in proposal 1, so we adjust the rolling tally

Proposal 1
yes: 100 - facuOldPower + facuNewPower (100 - 60 + 40 = 80)
no: 50
```

Does this sound ok to you?

Reply:
hmm i see what you mean, the best would be to take a snapshot of the weights when the proposal is created, but that runs into a similar issue as before. Maybe this is fine for now and in the future we can change it.